pbj-core/pbj-compiler/src/main/java/com/hedera/pbj/compiler/impl/Common.java

```diff
@@ -1,203 +1,688 @@
+        return fieldNumComment + cleanedComment;
+    }
+
+    /**
+     * Clean up a java doc style comment removing all the "*" etc.
+     *
+     * @param fieldComment raw Java doc style comment
+     * @return clean multi-line content of the comment
+     */
+    public static String cleanJavaDocComment(String fieldComment) {
+        return cleanDocStr(
+                fieldComment
+                        .replaceAll(
+                                "/\\*\\*[\n\r\s\t]*\\*[\t\s]*|[\n\r\s\t]*\\*/",
+                                "") // remove java doc
+                        .replaceAll("\n\s+\\*\s+", "\n") // remove indenting and *
+                        .replaceAll(
+                                "/\\*\\*", "") // remove indenting and /** at beginning of comment.
+                        .trim() // Remove leading and trailing spaces.
+                );
+    }
+
+    /**
+     * Clean a string so that it can be included in JavaDoc. Does things like replace unsupported
+     * HTML tags.
+     *
+     * @param docStr The string to clean
+     * @return cleaned output
+     */
+    public static String cleanDocStr(String docStr) {
+        return docStr.replaceAll("<(/?)tt>", "<$1code>") // tt tags are not supported in javadoc
+                .replaceAll(" < ", " &lt; ") // escape loose less than
+                .replaceAll(" > ", " &gt; ") // escape loose less than
+                .replaceAll(" & ", " &amp; ") // escape loose less than
+        ;
+    }
+
+    /**
+     * Convert a field type like "long" to the Java object wrapper type "Long", or pass though if
+     * not java primitive
+     *
+     * @param primitiveFieldType java field type like "int" etc
+     * @return java object wrapper type like "Integer" or pass though
+     */
+    public static String javaPrimitiveToObjectType(String primitiveFieldType) {
+        return switch (primitiveFieldType) {
+            case "boolean" -> "Boolean";
+            case "int" -> "Integer";
+            case "long" -> "Long";
+            case "float" -> "Float";
+            case "double" -> "Double";
+            default -> primitiveFieldType;
+        };
+    }
+
+    /**
+     * Recursively calculates the hashcode for a message fields.
+     *
+     * @param fields The fields of this object.
+     * @param generatedCodeSoFar The accumulated hash code so far.
+     * @return The generated code for getting the hashCode value.
+     */
+    public static String getFieldsHashCode(final List<Field> fields, String generatedCodeSoFar)
+            throws RuntimeException {
+        for (Field f : fields) {
+            if (f.parent() != null) {
+                final OneOfField oneOfField = f.parent();
+                generatedCodeSoFar += getFieldsHashCode(oneOfField.fields(), generatedCodeSoFar);
+            }
+
+            if (f.optionalValueType()) {
+                generatedCodeSoFar = getOptionalHashCodeGeneration(generatedCodeSoFar, f);
```

Comment:
This is questionable, to say the least. Key requirement to `hashCode()` implementation is forward compatibility. Whenever a new field is added to an entity class, it must not change its hash code. New fields are always optional, if they are included here, they would contribute to hash code value.

With that in mind, I can't say if all HS types we use today have some required fields (that can be used to calculate hash codes). If so, should all of them be accounted in `hashCode()`? If the answer to either of these questions is "no", there must be a different mechanism to specify the set of fields to be used for hash codes (annotations?)

Reply:
After talking to Jasper before, adding a new field actually should change the hashcode, if the value of an optional field is not null. In general, field values of null (whether optional or not) do not alter the hash code - default values (they are not considered in the calculation, except null in a list - a list with value noll at a certain position should change the dashcode, I think). Again, fields with def values for the corresponding type are not included in the dashcode calculation.